# 課題4.2（その１）
*****

学籍番号：

氏　　名：

*****

**■問題１：** まず次のセルを実行せよ。実行後、`pi_list` というリストが作成される。`pi_list` は、円周率 $\pi$ の小数第1000万位までの各１桁整数を順に要素にもつリストである。

In [1]:
import pickle
with open('pi_list.data', 'rb') as f:
    pi_list = pickle.load(f)

> `pi_list`は`[3,1,4,1,5,9,2,6,5,3,5,8,9,7,9,3,2,3,8,4,6,2,...]`というリストです。
>> ※）$\pi$ の任意の小数位までは、Python の外部モジュールである Sympy を使って得ることができます（Anaconda をインストールすると自動的にこのモジュールもインストールされます）。上のリストは，Sympy を使って、$\pi$ の小数1000万位までを得、リスト化し、pickle という標準モジュールを使ってバイナリデータにしたものです。すべてのことを Python で行なっています。

□1. `pi_list` の要素数を求めよ（教科書p.39参照）：

In [2]:
len(pi_list)

10000001

> 10000001 個ありますか？（整数部 3 はインデックス0の要素、小数第１位の 1 はインデックス1の要素、小数第２位の 4 はインデックス2の要素、小数第３位の 1 はインデックス3の要素、小数第４位の 5 はインデックス4の要素、... となるので、要素数はこうなるわけです。）

□2. `input`関数で1桁の数字を入力し、`for` と `if` 文を使って、$\pi$ の小数第1000万位までに `input`関数で入力した数字がいくつあるか求めるプログラムを作成せよ。ただし、プログラムを実行すると  
```
0～9の数字を入力：
```  
というプロンプトが表示され、例えば  
```
0～9の数字を入力：3
```  
と入力すれば、
```
3の個数 : 999965
```
と表示されるようにしておけ。

In [3]:
num = int(input('0～9の数字を入力：'))
counter = 0
for j in pi_list:
    if j == num:
        counter += 1
print(f'{num}の個数：{counter}')

7の個数：1000207


> ヒント：`counter` という変数を作りましょう（初期化のために `counter = 0` としておきます）。リスト `pi_list` から１つずつ要素を取り出し（← `for` 文）、取り出した数が 3 のとき（← `if` 文）、`counter` に 1 を足していきます。
>
> 流れ図は以下のようになります：
>
> <img src="pics/flowchart1.png">

□3．上のプログラムを実行（［Shift］＋［Enter］）し、`pi_list`に含まれる**3以外**の数字の個数を求めよ：  
*****

**■問題２：**`for`文と`range`関数を使って、**課題4.2**の**問題**のプログラムを作り直せ：

In [4]:
num = int(input('0～9の数字を入力：'))
counter = 0
for j in range(len(pi_list)):
    if pi_list[j] == num:
        counter += 1
print(f'{num}の個数：{counter}')

8の個数：999814
